### PR TITLE
Converts securitrons to mob critters

### PR DESCRIPTION
**Pull request by MylieDaniels**
[Critter][Code Quality]
## About the PR 
Converts securitrons to mob critters, with all the functionality therein.

I **highly** suggest this be testmerged for at least a week before a full merge is considered, because of the scope.
I'd like to resprite securitrons to better hold human-based inhands (they do that now), but that is out of scope!

The details:
- Securitrons are now a mob critter pathed at `/mob/living/critter/robotic/securitron`
image
- Securitrons now have a mob critter AI, pathed at `/datum/aiHolder/patroller/packet_based/securitron`.
- Securitrons have a radial menu when clicked on by someone with access with an empty hand on help intent, replacing the old Topic menu.
image
- Securitron batons contain the behavior for charging up and staying on for a few seconds, and this is preserved in the dropped Beepsky baton - might be a slight nerf to that.
- If you have access to a securitron's controls, it will not aggress on you for attacking it unless it is at 80% or less hp. Wearing a badge in addition moves this to 50% hp.
- Emagging, whistles, the PDA app, etc, it all still works from my couple days of testing.
- Admins have access to `Change hand item *` through admin interaction, which lets a securitron use approximately 99% of the items in the game as a baton replacement with functional AI.

There are two changes I implemented that could be atomized out:
- Securitrons without a unique name are now "Securitron-100", where 100 is a random 3 digit number, to allow for better tracking (probably going to be needed if players are ever spawned as them...)
image
- Securitrons now go off like a firecracker (explosion_new of power 3) if you emag them 6 times (which takes a full minute, with the cooldown on emagging them)- this is my reward to myself for this project.

There is also currently one feature removal:
- For now, they cannot be assembled by hand anymore- this is because I'd like to look over that with some developers and figure out how to untangle the mess that is assemblies.

## Why's this needed? 
It's time to give the people what they ~~want~~ deserve.
image

## Changelog 

```changelog
(u)MylieDaniels
(*)Securitrons are now mobs. Prepare thyself.
```

### CODE REVIEW

**Comment by Glamurio** [comment]
Obvious side-note, they should probably have the baton removed from their sprite, now that they actually hold one.

**Comment by MylieDaniels** [comment]
> Obvious side-note, they should probably have the baton removed from their sprite, now that they actually hold one.

Ofc ofc, though I want to resprite them in a separate, at least SLIGHTLY atomized PR.

**Comment by BatElite** [comment]
I know their timings won't be the same after this, but are they generally quicker or slower to react now?

**Comment by MylieDaniels** [comment]
> I know their timings won't be the same after this, but are they generally quicker or slower to react now?

I tried real hard to keep them as close to the same as I could, but without HEAVYWEIGHT_AI toggled on, they may be very very slightly slower to notice someone who is flagged with contraband/arrest status than before, but move more smoothly. All in all, I *think* I've got them very nearly the same speed, response time, etc. You can still juke them, can still sprint away from them, etc.

**Comment by frawhst** [comment]
error from the first live round that I saw
```
proc name: stack trace (/proc/stack_trace)
  source file: macros/debug.dm,6
  usr: (src)
  src: null
  call stack:
stack trace("Invalid A* start or destinatio...")
/datum/pathfind (/datum/pathfind): search()
get path to(Officer Beepsky (/mob/living/critter/robotic/securitron/beepsky), /list (/list), 150, null, 0, null, 1, null, 0, 1, null, 0)
moving (/datum/aiTask/succeedable/move): get path()
moving (/datum/aiTask/succeedable/move): on tick()
patrolling (/datum/aiTask/patrol/packet_based/securitron): on tick()
patrolling (/datum/aiTask/patrol/packet_based/securitron): tick()
/datum/aiHolder/patroller/pack... (/datum/aiHolder/patroller/packet_based/securitron): tick()
Mob AI (/datum/controller/process/mob_ai): doWork()
Mob AI (/datum/controller/process/mob_ai): doWork()
Mob AI (/datum/controller/process/mob_ai): process()
/datum/controller/processSched... (/datum/controller/processScheduler): runProcess(Mob AI (/datum/controller/process/mob_ai), 1.6)
```
a secbot did get destroyed, so they might be trying to path out of null for some reason?

**Comment by MylieDaniels** [comment]
Oh, AI isn't disposed when a mob is killed, is it.. will fix.

File: code/mob/living/critter.dm [hunk 1]
```diff
@@ -1034,7 +1039,8 @@ ADMIN_INTERACT_PROCS(/mob/living/critter, proc/modify_health, proc/admincmd_atta
 		handcount++
 		if (HH.object_for_inhand)
 			var/obj/item/I = new HH.object_for_inhand
-			var/image/inhand = image(icon = I.inhand_image_icon, icon_state = "[I.item_state][HH.suffix]",
+			var/suffix = I.two_handed ? "-LR" : "[HH.suffix]"
```
Comment (ZeWaka) [inline code/mob/living/critter.dm:1042]:
```suggestion
			var/suffix = I.two_handed ? "-LR" : HH.suffix
```
no need to stringify twice & later

File: code/mob/living/critter.dm [hunk 2]
```diff
@@ -1046,7 +1052,8 @@ ADMIN_INTERACT_PROCS(/mob/living/critter, proc/modify_health, proc/admincmd_atta
 				continue
 			if (!I.inhand_image)
 				I.inhand_image = image(I.inhand_image_icon, "", HH.render_layer)
-			I.inhand_image.icon_state = I.item_state ? "[I.item_state][HH.suffix]" : "[I.icon_state][HH.suffix]"
+			var/suffix = I.two_handed ? "-LR" : "[HH.suffix]"
```
Comment (ZeWaka) [inline code/mob/living/critter.dm:1055]:
```suggestion
			var/suffix = I.two_handed ? "-LR" : HH.suffix
```
x

File: code/mob/living/critter/ai/generic_critter.dm [hunk 3]
```diff
@@ -62,17 +62,20 @@
 	var/mob/living/critter/C = holder.owner
 	return C.can_critter_attack()
 
+/* Commented out and replaced with intent set inside critter attacks
```
Comment (ZeWaka) [inline code/mob/living/critter/ai/generic_critter.dm:65]: don't just comment out stuff

File: code/mob/living/critter/ai/generic_critter.dm [hunk 4]
```diff
@@ -116,7 +118,6 @@
 
 /datum/aiTask/sequence/goalbased/critter/attack/fixed_target/New(parentHolder, transTask, atom/fixed_target)
 	..(parentHolder, transTask)
-	add_task(holder.get_instance(/datum/aiTask/succeedable/critter/attack, list(holder)))
```
Comment (ZeWaka) [inline code/mob/living/critter/ai/generic_critter.dm:119]: ?

File: code/mob/living/critter/ai/shared.dm [hunk 5]
```diff
@@ -101,6 +101,21 @@
 	var/max_path_dist = 50 //keeping this low by default, but you can override it - see /datum/aiTask/sequence/goalbased/rally for details
 	var/list/found_path = null
 	var/turf/move_target = null
+	var/owner_id = null
+
+/datum/aiTask/succeedable/move/New()
+	. = ..()
+	if(isliving(src.holder.owner)) // todo: find out why the hell critters have "implants", which is where their implants are kept
+		var/mob/living/L = src.holder.owner // instead of in mob/living's implant list... and fix it
```
Comment (ZeWaka) [inline code/mob/living/critter/ai/shared.dm:109]: ?

File: code/modules/items/instruments/instruments.dm [hunk 6]
```diff
@@ -665,15 +665,16 @@ TYPEINFO(/obj/item/instrument/bikehorn/dramatic)
...
 	icon_state = "whistle-sec"
 	contraband = 4 //beepsky takes stolen whistles seriously
-	HELP_MESSAGE_OVERRIDE("Blow this to briefly command nearby securitrons to follow your pointing, point at a perp to have them arrested.")
+	HELP_MESSAGE_OVERRIDE("Blow this to briefly command a nearby securitron to follow your pointing, point at a perp to have them arrested.")
 
 	post_play_effect(mob/user)
 		var/list/bots = list()
-		for (var/obj/machinery/bot/secbot/secbot in view(user.client.view, user)) //cursed awful byond screen syntax
-			if (secbot.emagged)
+		for (var/mob/living/critter/robotic/securitron/secbot in view(user.client.view, user)) //cursed awful byond screen syntax
```
Comment (ZeWaka) [inline code/modules/items/instruments/instruments.dm:672]: viewers

File: code/obj/item/device/pda2/bot_control.dm [hunk 7]
```diff
@@ -87,40 +86,25 @@
 					. += "No bots found.<BR>"
 
 				else
-					for(var/obj/machinery/bot/secbot/B in src.botlist)
+					for(var/mob/living/critter/robotic/securitron/B in src.botlist)
 						. += "[B] at [get_area(B)]</A><BR>"
-						. += "Mode: "
-						switch(B.mode)
-							if(0)
```
Comment (ZeWaka) [inline code/obj/item/device/pda2/bot_control.dm:94]: no mode?

File: code/obj/item/device/pda2/bot_control.dm [hunk 8]
```diff
@@ -169,92 +153,96 @@
...
-				post_status("bot_control", "command", "bot_status", "active", active)
-				self_text("[active] action request sent.")
-
-			if("summonall")
+			if("summonall") // also not spoofable
 				if (!botlist.len)
 					PDA.updateSelfDialog()
 					return
 				SPAWN(0)
+					// trying this for real - mylie
```
Comment (ZeWaka) [inline code/obj/item/device/pda2/bot_control.dm:205]: ?

File: code/obj/item/device/pda2/bot_control.dm [hunk 9]
```diff
@@ -169,92 +153,96 @@
...
 	size = 8
 	header_thing = "Securitron Interlink PRO"
 	can_summon_all = 1
-	can_force_proc = 1
 
 /datum/computer/file/pda_program/bot_control/mulebot
 	name = "MULE Bot Control"
 	size = 16
 	var/list/beacons
+	var/obj/machinery/bot/mulebot/active
```
Comment (ZeWaka) [inline code/obj/item/device/pda2/bot_control.dm:245]: ?

File: code/modules/robotics/bot/disasters.dm [hunk 10]
```diff
@@ -1,24 +1,5 @@
...
-	desc = "A tattered and rusted security bot, held together only by the will of some wretched elder god."
-	health = 5
-	emagged = 1
-	auto_patrol = 1
-	no_camera = 1
-	var/blow_up = 1
-
-	New()
-		..()
-		src.name = pick("Commissar Beepevich","The Beeper","Murderbot","Killtron","Lawmaker")
```
Comment (ZeWaka) [inline code/modules/robotics/bot/disasters.dm:13]: removal?

File: code/modules/hydroponics/plants_herb.dm [hunk 11]
```diff
@@ -149,15 +149,10 @@ ABSTRACT_TYPE(/datum/plant/herb)
 
 	New()
 		. = ..()
-		START_TRACKING_CAT(TR_CAT_CANNABIS_OBJ_ITEMS)
```
Comment (ZeWaka) [inline code/modules/hydroponics/plants_herb.dm:152]: ?

File: code/mob/living/critter/mob_bots.dm [hunk 12]
```diff
@@ -399,3 +399,682 @@ ABSTRACT_TYPE(/datum/targetable/critter/bot/fill_with_chem)
...
+		src.UpdateName()
+	if(src.initial_limb)
+		var/datum/handHolder/HH = hands[1]
+		src.hud.add_object(src.equipped(), HUD_LAYER+2, HH.screenObj.screen_loc)
+
+	src.net_id = generate_net_id(src)
+
+	remove_lifeprocess(/datum/lifeprocess/blindness)
+	remove_lifeprocess(/datum/lifeprocess/blood)
+	remove_lifeprocess(/datum/lifeprocess/radiation)
```
Comment (ZeWaka) [inline code/mob/living/critter/mob_bots.dm:482]: why?

**Review by ZeWaka — COMMENTED** [review]
fixing the sprite to not have two batons is definitely part of the PR
making a unique inhand can wait until later

File: code/mob/living/critter/ai/generic_critter.dm (hunk 4 above)
Comment (MylieDaniels) [reply code/mob/living/critter/ai/generic_critter.dm:119]: The parent, /datum/aiTask/sequence/goalbased/critter/attack, does this as well- it's entirely unneeded and caused one extra attack past what seemed to be intended.

File: code/mob/living/critter/ai/shared.dm (hunk 5 above)
Comment (MylieDaniels) [reply code/mob/living/critter/ai/shared.dm:109]: I'm very confused by why mob/living/critter (which uses C.implants) has a separate handling of implants than mob/living (which uses L.implant) and want to fix it at a later date

File: code/modules/hydroponics/plants_herb.dm (hunk 11 above)
Comment (MylieDaniels) [reply code/modules/hydroponics/plants_herb.dm:152]: Ah crap, I must have accidentally undid this while removing any other testing on the patrol behavior (i used cannabis as the patrol beacons, easier to move than actual beacons)

File: code/mob/living/critter/mob_bots.dm (hunk 12 above)
Comment (MylieDaniels) [reply code/mob/living/critter/mob_bots.dm:482]: Beepsky dying to radiation is very unintended behavior imo, and reaching 100% rad resist seems to be impossible now

File: code/obj/item/device/pda2/bot_control.dm (hunk 7 above)
Comment (MylieDaniels) [reply code/obj/item/device/pda2/bot_control.dm:94]: I realized after writing the entire AI that this feature existed, and have no clue on how to reasonably rewrite it to support "what am I doing" reports

File: code/mob/living/critter/ai/generic_critter.dm (hunk 3 above)
Comment (MylieDaniels) [reply code/mob/living/critter/ai/generic_critter.dm:65]: Sure thing, will delete

File: code/obj/item/device/pda2/bot_control.dm (hunk 8 above)
Comment (MylieDaniels) [reply code/obj/item/device/pda2/bot_control.dm:205]: This used to be packetless, did not notice any performance issues with this version

**Pull request closed without merging**